language.md

```diff
@@ -7,312 +7,429 @@ the grammar of a synthesis problem, semantics for each production, and constrain
 
-#### Grammar Declarations
-All of the variables and non-terminals used in the grammar are declared first.
-```lisp
-  (declare-var (et et1 et2) E.Term)
-  (declare-var (bt bt1 bt2) B.Term)
+#### Semantics Declarations
 
-  (declare-var (x y r r1 r2) Int)
-  (declare-var (rb rb1 rb2) Bool)
-```
-A few notes about these variable declarations:
-* These are purely declarations, not bindings. The variable names are only being associated with a type.
-* Variables are universally quantified over the semantic CHCs, so variables do not store a state across productions.
-* Variables declared here, inside the grammar, are not visible outside the grammar.
+To associate semantic rules with a previously-declared syntax, we use the `define-funs-rec` command,
+which lets us declare semantic relations in the form of functions mapping values that the relations
+range over to Booleans. For example, the `E.Sem` relation expresses that the term `t_e: E` evaluates to
+`r` in a context given by `x` and `y`, and so we represent it as a function `(t_e, x, y, r) -> Bool`.
 
 ```lisp
-  (declare-nt E E.Term (E.Sem (E.Term Int Int Int)))
-  (declare-nt B B.Term (B.Sem (B.Term Int Int Bool)))
+(define-funs-rec
+  ((E.Sem ((t_e E) (x Int) (y Int) (r Int)) Bool)   ; Signature of sem. relation for E
+   (B.Sem ((t_b B) (x Int) (y Int) (rb Bool)) Bool) ; Signature of sem. relation for B
+   ...)
+  
+  (<E.Sem body> <B.Sem body> ...))
 ```
-The non-terminal declarations associate a non-terminal name (e.g., `E`) with the type of terms it produces (`E.Term`), as well as a
-semantic relation and signature (`E.Sem`). This relation is used in the semantics for productions, for some term produced by the 
-associated non-terminal. Note that this relation is also visible outside of the grammar and `synth-term` block, and it is used in
-the constraints.
 
-#### Non-terminals and CHC Heads
-After the grammar declarations, the non-terminals and associated productions are defined.
+Note that the bodies of the semantic relations should appear in the same order that their respective
+relations are defined in.
+
+Each semantic relation body should consist of a `match`-form that matches on the syntactic constructors
+for the term type. This way, we can assign to each constructor (corresponding to productions in the
+target grammar) a separate set of CHCs as semantic rules. The body of each match case corresponds to a
+single CHC. For terms with multiple semantic rules (for example, if-then-else conditionals or while
+loops), we may declare multiple bodies for a single match case.
+
 ```lisp
-((E et) (E.Sem et x y r)
-  <...productions...>)
-```
-The non-terminal and associated term variable are first, followed by the semantic relation for this non-terminal.
-This relation forms the head of all the CHCs formed by productions for this non-terminal:
-```
-\forall <...variables...>. <...CHC Body...> => (E.Sem et x y r)
+(match t_e
+  ($x <x semantic rule>)           ; Leaf term with no children

```

Comment:
I think there should be a list around all of the match clauses.

```lisp
(($x <x semantic rule>)
...
...))
```
The full example does properly have this in a list.